src/types.ts

```diff
@@ -1,24 +1,42 @@
-import { EventEmitter } from 'events'
+export type ExtractKey<T, U> =  { [K in keyof T]: T[K] extends U ? K : never }[keyof T]
+
 export interface Api {
   type: string
+  events: {
+    [key: string]: any
+  }
 }
 
-export abstract class API<T extends Api = any> {
-  public events = new EventEmitter()
+export type ApiListener<T> = (arg: T) => void
+
+export interface ApiEventEmitter<T extends Api> {
+  setMaxListeners(n: number): this
+  emit<K extends keyof T['events']>(type: K, arg: T['events'][K]): boolean
+  addListener<K extends keyof T['events']>(type: K, listener: ApiListener<T['events'][K]>): this
+  on<K extends keyof T['events']>(type: K, listener: ApiListener<T['events'][K]>): this
+  once<K extends keyof T['events']>(type: K, listener: ApiListener<T['events'][K]>): this
+  removeListener<K extends keyof T['events']>(type: K, listener: ApiListener<T['events'][K]>): this
+  removeAllListeners<K extends keyof T['events']>(type?: K): this
+  listeners<K extends keyof T['events']>(type: K): ApiListener<T['events'][K]>[]
+  listenerCount<K extends keyof T['events']>(type: K): number
+}
 
-  constructor(public type: T['type']) {}
+export type API<T extends Api> = {
+  type: T['type']
+  events?: ApiEventEmitter<T>
+  activate(): void
+  deactivate(): void
+} & {
+  [M in ExtractKey<T, Function>]: T[M]
 }
 
 export interface ModuleProfile<T extends Api = any> {
   type: T['type']
-  methods?: (Extract<keyof T, string>)[]
-  events?: (Extract<keyof T, string>)[]
+  methods?: ExtractKey<T, Function>[]
+  events?: (keyof T['events'])[]
 }
 
-export interface PluginProfile {
-  type: string
-  methods?: string[]
-  events?: string[]
+export interface PluginProfile<T extends Api = any> extends ModuleProfile<T> {

```

Comment:
`PluginProfile` refers to a plugin JSON declaration right?
why removing `type`, `methods`, and `events`?

Reply:
I'm not, it's extending the `ModuleProfile` instead which has the `type`, `methods` and `events`.